<a href="https://colab.research.google.com/github/AEIBIANCOFCE/AnalisisEstadisticoI/blob/main/3%20Muestreo/5_AEI_Muestreo_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<figure>
<center>
<img src='https://www.economicas.uba.ar/wp-content/uploads/2020/08/cropped-logo_FCE.png' />
</figure>

# **Universidad de Buenos Aires**
## **Facultad de Ciencias Económicas**

### **Análisis Estadístico I**

### Cátedra: Bianco

#### **Muestreo con Python**

***Material elaborado por la profesora Natalia Salaberry***

###**Librerías**

In [ ]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

###**Muestreo aleatorio Simple**

####**Con reposición**

Si de un conjunto de 100 números queremos obtener una muestra aletoria de 15 números con reemplazo, entonces realizamos lo siguiente

In [ ]:
#definimos un conjunto con los números del 1 al 100

conjunto = np.arange(1,101,1)

conjunto

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100])

In [ ]:
#tomamos la muestra de 15 números de modo aletorio con reemplazo

muestra_aleat_con_reemplzo = np.random.choice(conjunto,15,replace=True)

muestra_aleat_con_reemplzo

array([59, 20, 43, 59,  7, 47, 50, 87, 72, 15, 10, 30,  9, 33, 65])

In [ ]:
# si volvemos a ejecutar lo mismo:

muestra_aleat_con_reemplzo_1 = np.random.choice(conjunto,15,replace=True)

muestra_aleat_con_reemplzo_1

array([ 26,  67, 100,  22,  11,  29,   6,  37,  79,  89,  77,  62,  21,
        43,   5])

Observamso que los valores son diferentes.

Si se requiere mantener siempre la misma muestra entonces debemos setear una semilla

In [ ]:
np.random.seed(2024)

muestra_aleat_con_reemplzo_2 = np.random.choice(conjunto,15,replace=True)

muestra_aleat_con_reemplzo_2

array([  9,  97,   1,  28,  37,  63,  56,  42,  66,  61, 100,  75,  67,
        97,  38])

Si volvemos a ejecutar lo mismo con la misma semilla:

In [ ]:
np.random.seed(2024)

muestra_aleat_con_reemplzo_3 = np.random.choice(conjunto,15,replace=True)

muestra_aleat_con_reemplzo_3

array([  9,  97,   1,  28,  37,  63,  56,  42,  66,  61, 100,  75,  67,
        97,  38])

observamos que se obtienen los mismos valores

**En general, siempre se debe setear una semilla cuando se realiza muestreo aletorio**

Del mismo modo, podemos muestrear aleatoriamente sobre un DataFrame

In [ ]:
#Obtención y lectura de un DataFrame

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip

!unzip /content/bank.zip

data=pd.read_csv('/content/bank.csv', sep=';')

data

--2025-02-10 00:18:13--  https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘bank.zip’

bank.zip                [ <=>                ] 565.47K  3.46MB/s    in 0.2s    

2025-02-10 00:18:13 (3.46 MB/s) - ‘bank.zip’ saved [579043]

Archive:  /content/bank.zip
  inflating: bank-full.csv           
  inflating: bank-names.txt          
  inflating: bank.csv                


age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [ ]:
data.shape

(4521, 17)

Tenemos datos de 4521 clientes de un banco y vamos a seleccionar 1000 al aleatoriamente con reemplazo

In [ ]:
np.random.seed(2024)

clientes_1000_CR = data.sample(1000,replace=True)

clientes_1000_CR

age            job   marital  education default  balance housing loan  \
3578   53     technician   married  secondary      no      719     yes   no   
2656   37     technician   married  secondary      no      712     yes   no   
2688   27       services   married  secondary      no    -1202     yes   no   
2494   47         admin.  divorced  secondary      no      136      no  yes   
183    40         admin.  divorced  secondary      no        6      no   no   
...   ...            ...       ...        ...     ...      ...     ...  ...   
4449   31  self-employed    single  secondary      no      360     yes   no   
481    56     management    single  secondary      no      237     yes   no   
2472   54         admin.   married  secondary      no      865     yes   no   
770    26    blue-collar    single  secondary      no      -32     yes  yes   
1316   30     management   married   tertiary      no     3134     yes   no   

        contact  day month  duration  campaign  pdays  previous poutcome   y  
3578    unknown   14   may       230         1     -1         0  unknown  no  
2656   cellular   18   nov       291         2     -1         0  unknown  no  
2688    unknown    4   jun       749         2     -1         0  unknown  no  
2494   cellular   23   jul        70         1     -1         0  unknown  no  
183     unknown   11   jun       140         1     -1         0  unknown  no  
...         ...  ...   ...       ...       ...    ...       ...      ...  ..  
4449    unknown   26   may       592         4     -1         0  unknown  no  
481   telephone    9   jul       236         4     -1         0  unknown  no  
2472   cellular   17   nov        71         1     -1         0  unknown  no  
770    cellular   31   jul        14         6     -1         0  unknown  no  
1316   cellular   17   jul       124         1     -1         0  unknown  no  

[1000 rows x 17 columns]

####**Sin reposición**

Si de un conjunto de 100 números queremos obtener una muestra aletoria de 15 números con reemplazo, entonces realizamos lo siguiente

In [ ]:
#definimos un conjunto con los números del 1 al 100

conjunto = np.arange(1,101,1)

conjunto

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100])

In [ ]:
np.random.seed(2024)

muestra_aleat_con_reemplzo_0 = np.random.choice(conjunto,15,replace=False)

muestra_aleat_con_reemplzo_0

array([36, 20, 27, 18, 78, 25, 71, 59, 96, 15, 92, 88, 51, 50, 22])

Si volvemos a ejecutar lo mismo con la misma semilla:

In [ ]:
np.random.seed(2024)

muestra_aleat_con_reemplzo_00 = np.random.choice(conjunto,15,replace=False)

muestra_aleat_con_reemplzo_00

array([36, 20, 27, 18, 78, 25, 71, 59, 96, 15, 92, 88, 51, 50, 22])

observamos que se obtienen los mismos valores

**En general, siempre se debe setear una semilla cuando se realiza muestreo aletorio**

Del mismo modo, podemos muestrear aleatoriamente sobre un DataFrame

In [ ]:
#Obtención y lectura de un DataFrame

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip

!unzip /content/bank.zip

data=pd.read_csv('/content/bank.csv', sep=';')

data

--2025-02-10 00:18:13--  https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘bank.zip’

bank.zip                [ <=>                ] 565.47K  3.46MB/s    in 0.2s    

2025-02-10 00:18:13 (3.46 MB/s) - ‘bank.zip’ saved [579043]

Archive:  /content/bank.zip
  inflating: bank-full.csv           
  inflating: bank-names.txt          
  inflating: bank.csv                


age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [ ]:
data.shape

(4521, 17)

Tenemos datos de 4521 clientes de un banco y vamos a seleccionar 1000 al aleatoriamente con reemplazo

In [ ]:
np.random.seed(2024)

clientes_1000_CR = data.sample(1000,replace=False)

clientes_1000_CR

age           job   marital  education default  balance housing loan  \
3994   54  entrepreneur   married    unknown      no     1956      no   no   
4320   22    technician    single  secondary      no      333      no  yes   
1125   38   blue-collar  divorced  secondary      no     2998     yes   no   
4429   45       unknown   married    primary      no       44      no   no   
3036   73       retired   married    primary      no      279      no   no   
...   ...           ...       ...        ...     ...      ...     ...  ...   
1587   31        admin.    single  secondary      no     1306     yes   no   
3451   50    management    single    primary      no      766      no   no   
1203   48       retired   married    primary      no     5173     yes   no   
2965   35   blue-collar   married    primary      no     3587     yes   no   
803    57     housemaid    single  secondary      no      254      no   no   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
3994   cellular   19   nov       221         5     -1         0  unknown   no  
4320  telephone   12   nov       129         1     -1         0  unknown   no  
1125    unknown   11   jun        91         1     -1         0  unknown   no  
4429    unknown   11   jun        81         1     -1         0  unknown   no  
3036   cellular   28   jan       399         3     -1         0  unknown  yes  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
1587    unknown   28   may       217         5     -1         0  unknown   no  
3451    unknown   17   jun       174         2     -1         0  unknown   no  
1203    unknown   18   jun        88         8     -1         0  unknown   no  
2965    unknown   30   may       148         4     -1         0  unknown   no  
803    cellular    8   oct       676         1     -1         0  unknown   no  

[1000 rows x 17 columns]

###**Muestreo aletorio estratificado**

In [ ]:
#Se crea un Dataframe donde A son los estratos y B los valores de alguna variable
df = pd.DataFrame(dict(
    A=[1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4, 4],
    B=range(20)
))

df

A   B
0   1   0
1   1   1
2   1   2
3   1   3
4   1   4
5   1   5
6   1   6
7   2   7
8   2   8
9   2   9
10  2  10
11  2  11
12  2  12
13  3  13
14  3  14
15  4  15
16  4  16
17  4  17
18  4  18
19  4  19

In [ ]:
#se crea una muestra de tamaño n=5, estratificada por la columna A
muestra=df.sample(n=5, weights='A', random_state=1).reset_index(drop=True)
muestra

A   B
0  2  12
1  4  16
2  1   0
3  2  10
4  1   6

Lo verificamos:



In [ ]:
#Tamaño N
N=len(df)
N

20

In [ ]:
#observamos cuantos datos hay por estrato

df.groupby(df['A']).count()

B
A   
1  7
2  6
3  2
4  5

Hacemos la cuenta que propone el método, teniendo en cuenta la cantidad de casos por estrato:

\begin{equation}
\begin{array}{|c|c|}
    \hline
    A & B \\
    \hline
    1 & 7 \\
    \hline
    2 & 6 \\
    \hline
    3 & 2 \\
    \hline
    4 & 5 \\
    \hline
\end{array}
\end{equation}

Entonces:

Muestra en estrato 1:
\begin{align}
         n_1 = {5\over20}*7 = 1.75 \approx 2
\end{align}  
Muestra en estrato 2:
\begin{align}
         n_2 = {5\over20}*6 = 1.5 \approx 2
\end{align}

Muestra en estrato 3:
\begin{align}
         n_3 = {5\over20}*2 = 0.5 \approx 0
\end{align}

Muestra en estrato 4:
\begin{align}
         n_4 = {5\over20}*5 = 1.25 \approx 1
\end{align}

Siendo lo obtenido con Python anteriormente

Haciendo las mismas cuentas pero con Python:

In [ ]:
#redondeamos
round((5/N)*df.groupby(df['A']).count(),0)

B
A     
1  2.0
2  2.0
3  0.0
4  1.0

Esto nos indica que para generar una muestra de tamaño n=5, debemos tomar 2 casos del estrato 1, 2 casos del estrato 2, 0 casos del estrato 3 y 1 caso del estrato 4

Si por ejemplo, queremos aplicar a un DataFrame con mayor cantidad de variables, entonces:

In [ ]:
#Obtención y lectura de un DataFrame
#ya ejecutado al comienzo de este Colab
#!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip
#!unzip /content/bank.zip

data=pd.read_csv('/content/bank.csv', sep=';')
data

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

Supongamos que como variable que representan los estratos elegimos "education" y queremos una muestra de 1000 clientes por estratos.

Primero debemos convertir a categorías numéricas a "education" ya que para ser "Estrato" se requiere una variable numérica

In [ ]:
#observamos cantidad de clientes por categoría de "education"

data.groupby(data['education']).count()

age   job  marital  default  balance  housing  loan  contact  \
education                                                                  
primary     678   678      678      678      678      678   678      678   
secondary  2306  2306     2306     2306     2306     2306  2306     2306   
tertiary   1350  1350     1350     1350     1350     1350  1350     1350   
unknown     187   187      187      187      187      187   187      187   

            day  month  duration  campaign  pdays  previous  poutcome     y  
education                                                                    
primary     678    678       678       678    678       678       678   678  
secondary  2306   2306      2306      2306   2306      2306      2306  2306  
tertiary   1350   1350      1350      1350   1350      1350      1350  1350  
unknown     187    187       187       187    187       187       187   187

In [ ]:
#convertimos a categorías numéricas
data['education'] = data['education'].map({'primary':1,'secondary':2, 'tertiary':3, 'unknown':4})
data

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married          1      no     1787      no   no   
1      33       services  married          2      no     4789     yes  yes   
2      35     management   single          3      no     1350     yes   no   
3      30     management  married          3      no     1476     yes  yes   
4      59    blue-collar  married          2      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married          2      no     -333     yes   no   
4517   57  self-employed  married          3     yes    -3313     yes  yes   
4518   57     technician  married          2      no      295      no   no   
4519   28    blue-collar  married          2      no     1137      no   no   
4520   44   entrepreneur   single          3      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [ ]:
#generamos la muestra estratificada
muestra_data=data.sample(n=1000, weights='education', random_state=1).reset_index(drop=True)
muestra_data

age            job  marital  education default  balance housing loan  \
0     34     technician   single          2      no      440     yes   no   
1     31         admin.   single          2      no     1588     yes   no   
2     33       services  married          2      no     4789     yes  yes   
3     47       services  married          2      no     2597     yes   no   
4     45       services  married          2      no     1343      no  yes   
..   ...            ...      ...        ...     ...      ...     ...  ...   
995   30     technician   single          3      no     -970      no  yes   
996   34     technician  married          3      no     1539     yes   no   
997   53  self-employed  married          3      no     2627     yes  yes   
998   21        student   single          2      no     2488      no   no   
999   40    blue-collar  married          2      no     1100     yes   no   

      contact  day month  duration  campaign  pdays  previous poutcome    y  
0    cellular    6   may       445         2    345         2  failure   no  
1     unknown    3   jun        14         1     -1         0  unknown   no  
2    cellular   11   may       220         1    339         4  failure   no  
3     unknown   16   may       323         1     -1         0  unknown   no  
4    cellular    8   jul       220         1     -1         0  unknown   no  
..        ...  ...   ...       ...       ...    ...       ...      ...  ...  
995  cellular   21   aug       194         2     -1         0  unknown   no  
996  cellular   15   jun       441         1     56         1    other  yes  
997   unknown   16   jun        18        30     -1         0  unknown   no  
998  cellular   30   jun       258         6    169         3  success  yes  
999   unknown   29   may       660         2     -1         0  unknown   no  

[1000 rows x 17 columns]

In [ ]:
#observamos cantidad de clientes por estrato en la muestra
muestra_data.groupby(muestra_data['education']).count()

age  job  marital  default  balance  housing  loan  contact  day  \
education                                                                     
1           61   61       61       61       61       61    61       61   61   
2          471  471      471      471      471      471   471      471  471   
3          402  402      402      402      402      402   402      402  402   
4           66   66       66       66       66       66    66       66   66   

           month  duration  campaign  pdays  previous  poutcome    y  
education                                                             
1             61        61        61     61        61        61   61  
2            471       471       471    471       471       471  471  
3            402       402       402    402       402       402  402  
4             66        66        66     66        66        66   66

Se generó una muestra con 1000 casos conformados con 61 elementos en el estrato 1 ('primary'), 471 elementos en el  estrato 2 ('secondary'), 401 elementos en el  estrato 3 ('tertiary') y 66 elementos en el estrato 4 ('unknown')

###**Muestreo aleatorio por conglomerado**

In [ ]:
#Se crea un Dataframe
df = pd.DataFrame(dict(
    A=[1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4, 4],
    B=range(20)
))

df

A   B
0   1   0
1   1   1
2   1   2
3   1   3
4   1   4
5   1   5
6   1   6
7   2   7
8   2   8
9   2   9
10  2  10
11  2  11
12  2  12
13  3  13
14  3  14
15  4  15
16  4  16
17  4  17
18  4  18
19  4  19

In [ ]:
np.random.seed(2024)

# Número de conglomerados a seleccionar
n_conglomerados = 4  # Por ejemplo, seleccionamos 4 conglomerados al azar

# Seleccionar conglomerados aleatorios
conglomerados_seleccionados = np.random.choice(df['A'].unique(), n_conglomerados, replace=False)

# Fijar cantidad de elementos dentro de cada conglomerado
n_elementos_por_conglomerado = 5  # Por ejemplo, seleccionamos 5 elementos por conglomerado
#si tengo 20 elementos y especifiqué 4 conglomerados, quiero que haya 5 elementos por conglomerado

# Crear un DataFrame para almacenar las filas seleccionadas
muestra_2_etapas = pd.DataFrame()

# Para cada conglomerado seleccionado
for conglomerado in conglomerados_seleccionados:
    # Filtrar los elementos del conglomerado
    df_conglomerado = df[df['A'] == conglomerado]
    # Seleccionar una muestra aleatoria con reemplazo dentro del conglomerado
    muestra_conglomerado = df_conglomerado.sample(n=n_elementos_por_conglomerado, random_state=1, replace=True)
    # Agregar la muestra del conglomerado a la muestra total
    muestra_2_etapas = pd.concat([muestra_2_etapas, muestra_conglomerado])

# Observar la muestra final generada
muestra_2_etapas

A   B
12  2  12
10  2  10
11  2  11
7   2   7
8   2   8
18  4  18
19  4  19
15  4  15
16  4  16
18  4  18
14  3  14
14  3  14
13  3  13
13  3  13
14  3  14
5   1   5
3   1   3
4   1   4
0   1   0
1   1   1

De donde se observa que al haber muestreado con reemplazo hay filas más de una vez dentro de un conglomerado, como por ejemplo en el caso del conglomerado creado 4 (columa A), la fila 18; o en el conglomerado creado 3, la fila 13 y 14

Si por ejemplo, queremos aplicar a un DataFrame con mayor cantidad de variables, entonces:

In [ ]:
#Obtención y lectura de un DataFrame
#ya ejecutado al comienzo de este Colab
#!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip
#!unzip /content/bank.zip

data=pd.read_csv('/content/bank.csv', sep=';')
data

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [ ]:
np.random.seed(2024)

# Número de conglomerados a seleccionar
n_conglomerados = 10  # Por ejemplo, seleccionamos 10 conglomerados al azar

# Seleccionar conglomerados aleatorios
conglomerados_seleccionados = np.random.choice(data['age'].unique(), n_conglomerados, replace=False)

# Fijar cantidad de elementos dentro de cada conglomerado
n_elementos_por_conglomerado = 200  # Por ejemplo, seleccionamos 100 elementos por conglomerado por más que no abarquen el total
#si tengo 4521 elementos y especifiqué 10 conglomerados, quiero que haya 10 conglomerados con 200 elementos por conglomerado
#harán un toal de 2000, quednado 2521 elementos fuera de la muestra

# Crear un DataFrame para almacenar las filas seleccionadas
muestra_2_etapas = pd.DataFrame()

# Para cada conglomerado seleccionado
for conglomerado in conglomerados_seleccionados:
    # Filtrar los elementos del conglomerado
    df_conglomerado = data[data['age'] == conglomerado]
    # Seleccionar una muestra aleatoria con reemplazo dentro del conglomerado
    muestra_conglomerado = df_conglomerado.sample(n=n_elementos_por_conglomerado, random_state=1, replace=True)
    # Agregar la muestra del conglomerado a la muestra total
    muestra_2_etapas = pd.concat([muestra_2_etapas, muestra_conglomerado])

# Observar la muestra final generada
muestra_2_etapas

age         job   marital  education default  balance housing loan  \
3690   80   housemaid   married    primary      no        0      no   no   
2848   80     retired   married  secondary      no      462      no   no   
2896   80     retired   married    primary      no     1548      no   no   
688    80  management   married    primary      no     6483      no   no   
1312   80     retired   married  secondary      no     8304      no   no   
...   ...         ...       ...        ...     ...      ...     ...  ...   
3360   79     retired   married    primary      no      429      no   no   
3750   79     retired  divorced    unknown      no     2628      no   no   
3202   79     retired   married    primary      no     8556      no   no   
3202   79     retired   married    primary      no     8556      no   no   
3360   79     retired   married    primary      no      429      no   no   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
3690   cellular   23   feb       639         1    189         1  failure  yes  
2848   cellular    7   dec       127         2     -1         0  unknown   no  
2896   cellular    6   aug       331         2     -1         0  unknown   no  
688   telephone   19   oct       123         1     -1         0  unknown   no  
1312  telephone    9   jun       712         1     64        12  failure  yes  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
3360  telephone    8   jul      1006         2     -1         0  unknown  yes  
3750  telephone    8   jul       220         7    450         2  failure   no  
3202  telephone   18   jan       326         2    272         2  success  yes  
3202  telephone   18   jan       326         2    272         2  success  yes  
3360  telephone    8   jul      1006         2     -1         0  unknown  yes  

[2000 rows x 17 columns]

In [ ]:
muestra_2_etapas.groupby(muestra_2_etapas['age']).count()

job  marital  education  default  balance  housing  loan  contact  day  \
age                                                                           
22   200      200        200      200      200      200   200      200  200   
42   200      200        200      200      200      200   200      200  200   
46   200      200        200      200      200      200   200      200  200   
51   200      200        200      200      200      200   200      200  200   
59   200      200        200      200      200      200   200      200  200   
68   200      200        200      200      200      200   200      200  200   
73   200      200        200      200      200      200   200      200  200   
78   200      200        200      200      200      200   200      200  200   
79   200      200        200      200      200      200   200      200  200   
80   200      200        200      200      200      200   200      200  200   

     month  duration  campaign  pdays  previous  poutcome    y  
age                                                             
22     200       200       200    200       200       200  200  
42     200       200       200    200       200       200  200  
46     200       200       200    200       200       200  200  
51     200       200       200    200       200       200  200  
59     200       200       200    200       200       200  200  
68     200       200       200    200       200       200  200  
73     200       200       200    200       200       200  200  
78     200       200       200    200       200       200  200  
79     200       200       200    200       200       200  200  
80     200       200       200    200       200       200  200

In [ ]:
muestra_2_etapas.shape

(2000, 17)

Se obtuvo una muestra de 2000 elementos (filas), que se encuentra dividida en 10 conglomerados de edad

###**Muestreo sistemático**

In [ ]:
#Se crea un Dataframe
df = pd.DataFrame(dict(
    A=[1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4, 4],
    B=range(20)
))

df

A   B
0   1   0
1   1   1
2   1   2
3   1   3
4   1   4
5   1   5
6   1   6
7   2   7
8   2   8
9   2   9
10  2  10
11  2  11
12  2  12
13  3  13
14  3  14
15  4  15
16  4  16
17  4  17
18  4  18
19  4  19

In [ ]:
indexes = np.arange(0, len(df), step=5)
#arbitrariamente se indica tomar un valor cada 5 filas

systematic_sample = df.iloc[indexes]

systematic_sample


A   B
0   1   0
5   1   5
10  2  10
15  4  15

También lo podemos hacer sobre un DataFrame

In [ ]:
#Obtención y lectura de un DataFrame
#ya ejecutado al comienzo de este Colab
#!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip
#!unzip /content/bank.zip

data=pd.read_csv('/content/bank.csv', sep=';')
data

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [ ]:
indexes = np.arange(0, len(data), step=10)
#arbitrariamente se indica tomar un valor cada 10 filas

systematic_sample_data = data.iloc[indexes]

systematic_sample_data

age           job   marital  education default  balance housing loan  \
0      30    unemployed   married    primary      no     1787      no   no   
10     39      services   married  secondary      no     9374     yes   no   
20     38    management  divorced    unknown      no        0     yes   no   
30     68       retired  divorced  secondary      no     4189      no   no   
40     38    management    single   tertiary      no    11971     yes   no   
...   ...           ...       ...        ...     ...      ...     ...  ...   
4480   23   blue-collar   married    primary      no     1158     yes   no   
4490   58        admin.   married  secondary      no     3496     yes   no   
4500   38        admin.   married  secondary      no     4196     yes   no   
4510   36    technician  divorced  secondary      no      566     yes   no   
4520   44  entrepreneur    single   tertiary      no     1136     yes  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
0      cellular   19   oct        79         1     -1         0  unknown   no  
10      unknown   20   may       273         1     -1         0  unknown   no  
20     cellular   18   nov        96         2     -1         0  unknown   no  
30    telephone   14   jul       897         2     -1         0  unknown  yes  
40      unknown   17   nov       609         2    101         3  failure   no  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4480   cellular   16   apr       743         1     -1         0  unknown   no  
4490    unknown    2   jun       111         4     -1         0  unknown   no  
4500   cellular   12   may       193         2     -1         0  unknown   no  
4510    unknown   20   may       129         2     -1         0  unknown   no  
4520   cellular    3   apr       345         2    249         7    other   no  

[453 rows x 17 columns]

In [ ]:
systematic_sample_data.shape

(453, 17)

Se obtuvo una muestra de 453 elementos (filas)